src/umwm_advection.F90

```diff
@@ -165,6 +169,10 @@ subroutine refraction
                      * oneovdth
         ! integrate
         ef(o,p,i) = ef(o,p,i) - dtr * flux(o,p,i)
+        
+        if (fice(i) > fice_uth) then
+          ef(o,p,i) = 0.0
```

Comment:
refraction moves energy only in directional space so this check won't be necessary, as we're already zeroing out in advection

Reply:
I would have zeroed out the energy in umwm_run(), after propagation() and advection() are done, but I decided against that because umwm_run() is not used in GEOS, and it is one more thing to keep in mind when the GEOS code base is synced with the main repository.

There was no harm in zeroing out the energy in both  propagation() and advection(). That also makes it possible to have the full energy attenuation in cases when one may want to disable one or the other subroutine for testing/debugging purposes. But yes, I agree that in this this way there is some additional overhead that can be avoided.

Reply:
Roger that, I'm in agreement to keep zeroing out by ice in both subroutines.

As an aside, are we looking into applying seaice fraction as a window function for energy propagation, beyond simple zeroing out? An example of simplest filtering would be:

```
ef(:,:,i) = ef(:,:,i) * (1 - fice(i))
```

Rationale: If you have some seaice below the upper threshold, say 0.2, then you'd let 80% of energy to go through. 

Maybe this belongs to "level 2" of seaice implementation, where level 1 is simple zeroing out based on threshold. Level 3 could filter by frequencies -- shorter waves being more heavily filtered than long waves -- you get the idea. Just musing.

Reply:
I see, we're applying `(1 - fice(i))` to both Sin and Sds calculation. Interesting.

For Sin, I think this can be... OK as a first-order approach. This means in 10% seaice fraction, we will grow 90% of wave energy than we would otherwise.

However, for Sds, this is almost certainly wrong -- dissipation of wave energy should not be reduced because there is sea-ice. If anything it should be increased, which is what Kohoun's `s_ice` does.

I will remove `sds(o,p,i) = (1 - fice(i)) * sds(o,p,i)` from `sds_d12`. I will leave the same in `sin_d12`. I think this one is reasonable, at least on the first order.

Reply:
> But yes, I agree that in this way there is some additional overhead that can be avoided.

propagation and refraction take up a relatively small fraction of compute time relative to source functions, so I am not too concerned about this overhead.

Reply:
I will start with the most basic question: What quantities does the model predict in presence of ice?
If we focus on the wave energy I would say that we have two options: we limit the WM to operate over the open water fraction of a grid cell OR we introduce _grid cell mean energy_. The current implementation uses the latter. I define the mean grid cell energy as the area weighted wave energy in the fractional area free of sea ice influence/effects and the wave energy in the fractional area 'feeling' the presence of sea ice, that is:   _E = (1 - f)e + fe'_, where e is the wave energy over open water, _e'_ is the energy associated with the fractional area affected by the sea ice and _f_ is the 'effective' sea ice fraction in the grid cell. I can make some arguments that  _f_ should be larger than the actual sea ice fraction _f_ice_.  But what is _e'_? The very simple way to think about that is that if the ice fraction gets sufficiently large we 'put a lid' on the waves, so _e'_ could be 0. A more realistic treatment is perhaps to assign the high frequency tail of the spectrum to _e'_!? In any case I would consider _e'_ to be time invariant.

With that being said, lets see what all that means.  The time derivative of _E_ is: dE/dt  = (1-f)de/dt + (e'-e)df/dt_ 
If we look at the second RHS term we see that this is the change associated with expanding/contracting the ice fraction. An example that should help with interpreting that term would be a grid cell where the some of the new ice is formed so the energy in that additional sea-ice area would change from e to e' and hence the grid mean energy would be affected even if de/dt =0. I am not an sea ice expert, but I would expect that at the time scales typically used to do time integration df/dt << 1, so it is probably fine to neglect that term. Hence we can approximate _dE/dt ~= (1-f)de/dt_.  

My interpretation of the last equation is that to represent the effects of sea ice on the mean grid cell wave energy one needs to scale the source functions with (1-f) factor. Of course one has to also introduce a new term describing the actual attenuation of the  mean grid cell energy E.

...I need to update the documentation with a section about the attenuation of waves from sea ice.

Reply:
Got it, I can follow your argument. I do think there's a scale issue that's becoming more clear after reading your comment.

If we use grid cell mean energy, I think your math works out. In that case, you don't need to scale the source functions with (1-f), but only the time integrand: E(n+1) = E(n) + (1-f) * dE.

However, I have issue with grid cell mean energy. Take this situation for example: A grid cell with half solid ice, half ice-free open water. Seaice fraction is 0.5. Over ice, we have no waves, so energy there is zero. Over open water, we have wave energy such that Hs = 1 m. If you stick a buoy in the water, it will measure 1 m. Stick it in the ice, it will measure 0 m. Now take grid cell mean energy, now exactly half of the ice-free wave energy. The model reports Hs of ~0.7 m now. However, that's the wrong answer for either the ice-covered area, or the ice-free area. 

Of course, point measurements can't be exactly compared to grid-cell representative value from the model. Except that that's all we got and do it all the time. In fact, all the source functions (including Kohoun et al 2014, and their Hs = 3 m threshold) are derived from point measurements, and operate on scales of individual waves (10**-2 - 10**2 m). On these scales, there's either ice or no ice, but rarely something in between. This is why I believe it's not correct to apply seaice fraction (scales of 10**3 - 10**5 m) to processes that work on scales of individual waves.

Another challenge is getting clear on the meaning of seaice fraction. The situation in the example above is different from a situation where we have a uniform icy slush that is exactly half ice half water. How would we deal with seaice that is more like slush? Would we need another kind of seaice fraction?

I do see value in grid cell mean energy, but for diagnostic purposes, such as computing atmosphere or ocean stresses, which are by definition grid-cell averaged.

Reply:
Agreed, this is not an approach that can be expected to work well in all cases. But we have to keep in mind that right now we are very limited in terms of how sea ice is described and quantified - we have only sea ice fraction. We do not have information about ice type, thickness, etc. Without more detailed description of sea ice, the alternative is to stick with a simple threshold approach where the wave energy is completely attenuated when/where the ice fraction exceeds some magical threshold value (effect of ice is ignored otherwise). OR try to bring in a more processes/physics and hope that this will improve the model representativeness and skill. Ultimately, comparison with observations will answer which one of these approaches is better.

My hope is that the modulation of the source terms, the new parametrization of wave attenuation in the marginal ice zone, and the complete attenuation of waves when/where the ice fraction is close to 1 would work in tandem and give more reasonable waves.

Reply:
> Without more detailed description of sea ice, the alternative is to stick with a simple threshold approach where the wave energy is completely attenuated when/where the ice fraction exceeds some magical threshold value (effect of ice is ignored otherwise). 

How about applying seaice fraction (f) to the wave propagation flux? To me, this seems as the natural first step. If we have a cell with 0.5 seaice fraction (think 0.5 solid obstruction), approx. 0.5 energy should propagate through that cell, analogous to obstruction grids approach, except that we don't have the seaice grid. The special case of this is complete attenuation where seaice fraction is 1. 

S_ice is a good second step, and the challenge is determining what S_ice should be.

Reply:
The obstruction approach attempts to account for the location of the unresolved sub-grid obstruction elements, and it can do that because these locations are known. The idea is that two or more obstruction elements along the line of wave propagation would block as effectively as a single element. Given the spatial scales we run the wave model at, I would not worry so much about solid ice floats with sizes of 10--50 km. if we are talking about that big ice blocks, we are probably close to grid cells that are completely covered by ice, and waves that reach that location would be likelly suppressed by one of the neighboring cells. If we are talking about smaller randomly placed (and sufficiently thick) ice floats the obstruction should be close to 100% (it is like being in the forest and seeing only overlapping tree trunks unless you are near the edge of the forest). I would expect that what we have already put together to work reasonably well in such cases.

Reply:
Sure, I agree with this. I'm using the obstruction grids more as an analogy to describe the kind of filtering I'm talking about. Currently I see your zeroing-out of energy in propagation as a special case of this filtering, where the filter function is a step function such that ice flux is 1 for `fice < fice_uth` and 0 for `fice > fice_lth`.

As you say, this **may** work reasonably well. However, I'd say you need a good reason to choose a hard, discrete limit. Where does `fice_uth` come from? Fit to some data? Fine, but we know that physics of continuous fields don't operate with discrete thresholds. So even if we know there is some magical number around 0.7, why don't we make a continuous (say, sigmoid) function centered on that value that would mimic the behavior but not introduce discrete artifacts?

This is what I suggest:

* Isolate effects of sea-ice in their own source terms, S_ice*. `fice` should not percolate into other source functions that model ice-free physics.
* There may be some overlap / double counting of processes between current S_ice (Kohoun et al. 2014), and discrete zeroing out (rationale: If we know that most wave energy should be dissipated in `fice > 0.7`, then a correct source term should already be dissipating most energy in `fice > 0.7`).
* Allow finer control of these effects, a-la WW3. I think we should have distinct namelist switches for Sice_k14 (Kohoun et al, 2014) and ice filtering in advection.
* [WW3 manual](https://polar.ncep.noaa.gov/waves/wavewatch/manual.v5.16.pdf) is a good summary of what has been already done with seaice and waves.

Reply:
> How about applying seaice fraction (f) to the wave propagation flux? 

I misunderstood your suggestion. I read it as - lets only attenuate energy from ice in propagation and remove S_ice, etc.

> Where does `fice_uth` come from? Fit to some data? Fine, but we know that physics of continuous fields don't operate with discrete thresholds. So even if we know there is some magical number around 0.7, why don't we make a continuous (say, sigmoid) function centered on that value that would mimic the behavior but not introduce discrete artifacts?

Isn't step function already used to deal with shallow water and shore line. After all a mask is created based on some threshold criteria and energy is set to zero for points with depth < 10m. I do not see much of difference between doing that and setting e=0 here and there based on a dynamic ice mask. In principle I could have used the 'land-water' mask and extend it to 'land-ice-water' mask by adding another criteria related to the ice fraction. In practice this is not going to work well without tinkering with the code, because the mask would be dynamic and the spatial dimension of the unrolled arrays will change at every time step and you will have to deal with that (allocate and deallocate memory at every time step, etc).

In regard to using S-function... I have no objections. We still have to come up of with two parameters to describe the function. S-function may very well work better than a step function at fine resolutions. I am not sure how much of a difference it would make at resolutions 10--50km because of the strong gradient in the ice fraction. 

>     * Isolate effects of sea-ice in their own source terms, S_ice*. `fice` should not percolate into other source functions that model ice-free physics.

Could you please elaborate on this one. Did we not agree earlier that it make sense to reduce the wind input source term in presence of ice. How would you do that without using some information about the ice field? 

>     * There may be some overlap / double counting of processes between current S_ice (Kohoun et al. 2014), and discrete zeroing out (rationale: If we know that most wave energy should be dissipated in `fice > 0.7`, then a correct source term should already be dissipating most energy in `fice > 0.7`).

That double counting should be harmless (0*0 = 0). The very mundane and practical reason to set e=0 above certain ice fraction is the fact that diagnostics is done after propagation and refraction and it will be very strange to see non-zero energy written out where it is supposed to be zero. 

That brings another question about what should be the diagnostics in locations with large ice fraction. To me it makes sense to set diagnostics to undefined/missing value, similarly to what is done over land. In order to do that we again have to employ some threshold value for the ice fraction.

Reply:
> Isn't step function already used to deal with shallow water and shore line. 

Where we have water we solve for waves, where we have land, we don't. While I see where you're going with the analogy, I don't think it's comparable. In contrast to seamask, with seaice we're jumping from all waves in `fice == 0.69` to no waves in `fice == 0.7`. The big difference is that we have much better data on land, and we have much better knowledge about water waves over dry land. Seaice fraction is less known, more dynamic, and we don't understand it as well as dry land.

On the other hand, I do like that for a hard threshold we have to pick only one number, as you say, and a more complex function would need two or more.

> After all a mask is created based on some threshold criteria and energy is set to zero for points with depth < 10m

No, we limit the depth to `dmin` so prevent very shallow waters that would cause extremely short time steps, but we don't zero out wave energy there.

All that being said, the model does have a few other magical threshold values. I'm not hard fast against these, I'm just reluctant to introduce even more into the model. But considering that we're building this to be easily controlled via namelist parameters, I think it's a good first step to go with a step function, in its own source term (or even the existing Sice).

> Could you please elaborate on this one. Did we not agree earlier that it make sense to reduce the wind input source term in presence of ice. How would you do that without using some information about the ice field?

After your explanation that modulating existing source terms comes from reframing the prognostic variable of the model to the "grid cell mean energy", I don't see why the ice-free source functions (Sin, Sds, Snl, Sdt, Sbot) should be modulated by presence of ice. Instead, wave energy should be modulated by seaice through its respective source terms -- Sice and ice-filtering of wave energy. I do think that grid cell mean energy will be a useful diagnostic.

The way that sea-ice effects will implicitly get into the other source terms is through the change in water properties:

* Water density
* Surface tension
* Water viscosity

This will affect at least Sin, Sds, and the dispersion relationship. Obviously, this needs more data.

Reply:
I am not nit picking, but...

> Where we have water we solve for waves, where we have land, we don't.

The wave energy is set to 0 over 'land' points which also happen to include shallow coastal water points. And one obviously wants to do that for several reasons: 1) dissipate waves at the  coast line (reflection of waves is not considered) ; 2) wave propagation requires energy for the neighbouring points which may happen to be 'land' points; 3) parametrization of physics/processes considered in UMWM. How is that different than setting e=0 but for points with ice fraction larger than some value? 

> Instead, wave energy should be modulated by seaice through its respective source terms -- Sice and ice-filtering of wave energy.

I do not know how to do that.
 
> The way that sea-ice effects will implicitly get into the other source terms is through the change in water properties:
> 
>     * Water density
> 
>     * Surface tension
> 
>     * Water viscosity

What is the argument for ice to change water density, tension and viscosity used in the source terms if we are going to apply them only over open water? That seems inconsistent to me.

Reply:
> The wave energy is set to 0 over 'land' points which also happen to include shallow coastal water points.

Can you point me to where we're doing this in the code? From the top of my head I believe we don't do this but maybe I'm forgetting something.

>> Instead, wave energy should be modulated by seaice through its respective source terms -- Sice and ice-filtering of wave energy.
>
> I do not know how to do that.

You already did :) -- subroutine `s_ice` and zeroing out of wave energy in propagation and refraction.

> What is the argument for ice to change water density, tension and viscosity used in the source terms if we are going to apply them only over open water? That seems inconsistent to me.

Sorry for the confusion -- these effects apply only where we have ice smaller than < O(100 m). In freshly forming ice, waves break the ice sheets and ice effectively modulates water density, viscosity, and surface tension. I can't think of a way to incorporate this effects with seaice fraction alone.

Reply:
> > The wave energy is set to 0 over 'land' points which also happen to include shallow coastal water points.
> 
> Can you point me to where we're doing this in the code? From the top of my head I believe we don't do this but maybe I'm forgetting something.

* umwm_init.F90, gitlab version

! Allocate and add a ghost land point at the beggining: 
ALLOCATE(e(om,pm,iistart-1:iiend))
ALLOCATE(cp0(om,iistart-1:iiend))
ALLOCATE(cg0(om,iistart-1:iiend))
ALLOCATE(k(om,istart:iend))

! Initialize:
e = TINY(e)
...
! Treat land points for cp0, cg0, and e:
! (needed for advection terms)
DO i=istart,iend
  IF(is(i) < iistart .OR. is(i) > iiend)is(i) = iistart-1
  IF(in(i) < iistart .OR. in(i) > iiend)in(i) = iistart-1
  IF(ie(i) < iistart .OR. ie(i) > iiend)ie(i) = iistart-1
  IF(iw(i) < iistart .OR. iw(i) > iiend)iw(i) = iistart-1
ENDDO

If that was changed in the github code then I take my words back.

Reply:
OK, got it, I see what you mean. We still have this. What this does is it makes sure that the advective flux from a land cell into the water cell is always zero. The remaining difference is just semantics so I'm happy to move on. :)

Let's stick with a step function for now. I like the idea of a single tunable parameter here (`fice_uth`).

Do you agree with:

* Moving `if (fice(i) > fice_uth) ef(o,p,i) = 0` into its own subroutine? This could then be called from `umwm_run`. My rationale for this separation is that zeroing out energy in high seaice is neither an advective nor a refractive process.
* `(1 - fice(i))` should not modulate Sin or Sds, assuming we're modeling open-water waves with these terms?

Reply:
>     * Moving `if (fice(i) > fice_uth) ef(o,p,i) = 0` into its own subroutine? This could then be called from `umwm_run`. My rationale for this separation is that zeroing out energy in high seaice is neither an advective nor a refractive process.

Sure, that is fine.



>     * `(1 - fice(i))` should not modulate Sin or Sds, assuming we're modeling open-water waves with these terms?

If we model only waves in the open-fraction of the cell, what wave energy should be used and attenuated in S_ice then?

Reply:
> If we model only waves in the open-fraction of the cell, what wave energy should be used and attenuated in S_ice then?

In my view, open-water wave energy (as currently modeled by UMWM without ice) is the only meaningful quantity to attenuate by ice.

Picture: ice-free waves (`fice = 0`) propagate toward a grid cell with ice say, `fice = 0.5`. Once they enter, we want to know how much should wave energy decrease due to effects of ice. All of the energy that enters is subject to ice effects. In contrast, if we applied Sice to `(1 - fice) * e`, we'd decrease only half of the wave energy, unless you redefine the wave energy to be something else in presence of ice. I don't think it should, as its a per unit area quantity, analagous to surface fluxes in atm. and ocean circulation models.

Reply:
In regard to the question whether the source terms should be modulated in presence of ice I will point you to a recent article (http://dx.doi.org/10.1002/2015GL064715) that validates the sea ice physics parameterizations in WW3 with observations. Here is the relevant excerpt from the Conclusions section:
>  WW3 currently treats the wind input, nonlinear interaction, and turbulent dissipation terms in partially ice-covered seas the same as in open water. The only adjustment is to multiply each term by the fraction of open water in ice-covered zones.

Reply:
Thank you for this paper Anton, it looks like it will be very useful and fortunately it is short enough, so I will definitely read it. :)

> The only adjustment is to multiply each term by the fraction of open water in ice-covered zones.

I can't say that this makes sense to me yet, but I may very well be wrong. I will read the paper and sit with this.

Reply:
This one is a good review on the subject: https://royalsocietypublishing.org/doi/full/10.1098/rsta.2017.0342 .

Reply:
> In regard to the question whether the source terms should be modulated in presence of ice I will point you to a recent article (http://dx.doi.org/10.1002/2015GL064715) that validates the sea ice physics parameterizations in WW3 with observations. Here is the relevant excerpt from the Conclusions section:
> 
> > WW3 currently treats the wind input, nonlinear interaction, and turbulent dissipation terms in partially ice-covered seas the same as in open water. The only adjustment is to multiply each term by the fraction of open water in ice-covered zones.

From the documentation of WW3 version 4.18 (Tolman et al. 2014, page 14):
>  The treatment of partial ice coverage (ice concentration) in the source term follows the concept of a limited air-sea interface. This means that the momentum transferred from the atmosphere to the aves is limited. Therefore, input and dissipation terms are scaled by the fraction of ice concentration. The nonlinear wave-wave interaction term can be used in areas of open waterand ice (Polnikov and Lavrenov, 2007). The scaling is implemented so that it is independent of the source term selected.

Reply:
Do you understand the physical meaning of this approach? I can't seem to, and everything I read makes it seem like it should be trivial. Why are there no explanations for this? To me it's counterintuitive. 

Limited air-sea interface I can get my head around for wind input. There some ocean surface covered with ice and wind can't act there to generate waves. Within a single step and a single grid cell, net wave growth is a cumulative effect of many small and short waves building up over the time step. It's not obvious to me why the ice-effect is exactly linear with ice fraction, but, OK. I can swallow it. 

Where I get really tripped up is the wave dissipation. Take for example a wave packet with energy E moving into a cell with ice fraction of 0.5. For simplicity, let's assume that wind speed is exactly the phase speed of the wave so that Sin = 0, and that it takes a unit of time for the waves to move through the cell. Let's also assume that wave dissipation Sds is also 1 for simplicity.

How much wave energy comes out of the icy cell on the other side? 

E(n+1) = E(n) * exp[-Sds * E(n)] ~= 0.37

or 

E(n+1) = E(n) * exp[-Sds * (1 - fice) * E(n)] ~= 0.61

Why is it that in presence of ice, same wave energy E should be dissipated less than in ice-free water?

Of course, I omitted Sice in the calculation here. The only counter-argument I can imagine is that in presence of ice Sds should decrease so that Sice can take over. But this implies that somehow Sice can compensate for the linear scaling of Sds with `fice`, and dissipate some more due to actual ice-wave interaction.

Assuming we accept that Sin and Sds should be scaled with `(1 - fice)`. How about Sdt, Sdv, and Sbot? Should they be too? How about Sice?